### PR TITLE
Refactor analyzeTree to use ls-tree

### PR DESCRIPTION
**Pull request by joglr**
This PR uses ls-tree to get the total structure and size of blobs in the tree
It also queues up parseBlame calls and uses Promise.all to not waste time.
git-truck no longer reports amount of lines in a file, instead amount of bytes are reported.

Benchmarks run on my XPS with git-truck as target repository:

```
Main
6.67s
7.09s
6.99s

ls-tree
4.49s
4.38s

ls-tree + Promise.all for parseBlame (this PR)
1.35s
1.26s
1.15s
1.21s
```

### CODE REVIEW

File: src/analyzer/analyze.server.ts [hunk 1]
```diff
@@ -79,48 +79,84 @@ export async function analyzeCommit(repoName: string, hash: string): Promise<Git
...
+    name,
+    hash,
+    children: [],
+  } as GitTreeObject
+
+  const jobs = []
+
+  for (const child of lsTreeEntries) {
+    const prevTrees = child.path.split("/")
+    const newName = prevTrees.pop() as string
```
[the remark this answers is not in the document]
Comment (joglr) [reply src/analyzer/analyze.server.ts:121]: @tjomson

The split operation is guaranteed to produce an array with length > 0
If there is no split, it produces an array with the entire string
```
"asd".split("/") -> ['asd']
```
I would argue this is safe.

File: src/analyzer/analyze.server.ts [hunk 2]
```diff
@@ -79,48 +79,84 @@ export async function analyzeCommit(repoName: string, hash: string): Promise<Git
...
+    const newPath = `${path}/${child.path}`
+    let currTree = rootTree
+    for (const treePath of prevTrees) {
+      currTree = currTree.children.find((t) => t.name === treePath && t.type === "tree") as GitTreeObject
```
[the remark this answers is not in the document]
Comment (joglr) [reply src/analyzer/analyze.server.ts:125]: I can't really see how to do this in a prettier way. We iteratively go down each path segment in the tree and lookup the next segment in the previous tree.

For example:
``typescript
newPath = "git-truck/src/analyzer/constants.ts"

const prevTrees = ["src", "analyzer", "constants.ts"]
prevTrees.pop() // prevTrees = ["src", "analyzer"]

currTree = 
{
  type: "tree",
  name: "git-truck",
  path: "git-truck",
  children: [
    {
      type: "tree",
      name: "src",
      path: "git-truck/src",
      children: [
        {
          type: "tree",
          name: "analyzer",
          path: "git-truck/src/analyzer",
          children: [
            {
              type: "blob",
              path: "git-truck/src/analyzer/constants.ts",
              name: "coauthors.server.ts",
            },
          ],
        },
      ],
    },
  ],
}

for (const treePath of prevTrees) 
  1. iteration
  currTree = 
  {
    type: "tree",
    name: "src",
    path: "git-truck/src",
    children: [
      {
        type: "tree",
        name: "analyzer",
        path: "git-truck/src/analyzer",
        children: [
          {
            type: "blob",
            path: "git-truck/src/analyzer/constants.ts",
            name: "coauthors.server.ts",
          },
        ],
      },
    ],
  },
  2. iteration
  currTree = 
  {
    type: "tree",
    name: "analyzer",
    path: "git-truck/src/analyzer",
    children: [
      {
        type: "blob",
        path: "git-truck/src/analyzer/constants.ts",
        name: "coauthors.server.ts",
      },
    ],
  },
```

File: src/analyzer/analyze.server.ts [hunk 3]
```diff
@@ -79,48 +79,84 @@ export async function analyzeCommit(repoName: string, hash: string): Promise<Git
...
-          name,
-          content: await GitCaller.getInstance().catFileCached(hash),
-          blameAuthors: await GitCaller.getInstance().parseBlame(newPath),
-        })
+          name: newName,
+          sizeInBytes: child.size as number,
+          blameAuthors: {}
+        }
+        jobs.push((async (path) => {
+          blob.blameAuthors = await GitCaller.getInstance().parseBlame(path)
```
[the remark this answers is not in the document]
Comment (joglr) [reply src/analyzer/analyze.server.ts:150]: It has to be a function, such that we can await and assign the blameAuthors later, so we don't block the current loop. 

Or if it has to be a function, then something like

> ```
> (async () => blob.blameAuthors = await GitCaller.getInstance().parseBlame(blob.path))
> ```
> ?

Good suggestion, this works as well and is much simpler. For some reason, I couldn't get something like that to work when I implemented it.

File: src/analyzer/model.ts [hunk 4]
```diff
@@ -67,21 +67,21 @@ export interface GitBlobObject extends GitBaseObject {
...
 export interface HydratedGitBlobObject extends GitBlobObject {
-  noLines: number
   authors: Record<string, number>
   noCommits: number
   lastChangeEpoch?: number
   isBinary?: boolean
   unionedAuthors?: Map<AuthorshipType, Record<string, number>>
   dominantAuthor?: Map<AuthorshipType, [string, number]>
   isSearchResult?: boolean
+  blameAuthors: Record<string, number>
```
[the remark this answers is not in the document]
Comment (joglr) [reply src/analyzer/model.ts:84]: Removing

**Review by tjomson — APPROVED** [review]
🥌

**Pull request merged**